src/python/pants/rules/core/binary.py

```diff
@@ -7,18 +7,21 @@
 from pants.engine.addressable import BuildFileAddresses
 from pants.engine.console import Console
 from pants.engine.fs import Digest, DirectoryToMaterialize, Workspace
-from pants.engine.goal import Goal, LineOriented
+from pants.engine.goal import Goal, GoalSubsystem, LineOriented
 from pants.engine.legacy.graph import HydratedTarget
 from pants.engine.rules import console_rule, rule, union
 from pants.engine.selectors import Get, MultiGet
 
 
-@dataclass(frozen=True)
-class Binary(LineOriented, Goal):
+class BinaryOptions(LineOriented, GoalSubsystem):
   """Create a runnable binary."""
   name = 'binary'
 
 
```

Comment:
This feels pretty unfortunate. I would almost rather that we did some sort of MyPy hackery than require this kind of boilerplate.

Reply:
There are two cost axes here afaict:
1. boilerplate
2. non-uniformity

The less boilerplate solution is nice for typing. It's also nice if you don't think about things too much.

The minute you do think about things though, this breaks with the learned patterns in all other rules - options are registered on Subsystem classes. You can't use a smart editor to track types here for a big example. Fancy editor defeats aside, the same brain cells are defeated.

It would be nice to find magic to get both but no one can seem to do this.
As such it seems we should decide which is more important - making it slightly simpler to type one rule type out. Or making Goal work like every other rule type.

I think I vote for the latter. I've been dumbfounded by the current magic on more than one occasion.

Reply:
I tend to vote for the latter as well. On balance this seems like too much magic for too little boilerplate savings.  

Apart from the mypy question, the current method is pretty unintuitive to rule authors: You have a `register_options` method on a thing that isn't an Optionable, which goes against usual expectations. And then an Optionable you have no way of knowing about from introspection gets created for you, and you have to "just know" about it. It's not ideal. 

So the boilerplate can be said to have an advantage - in that it telegraphs to the reader what's actually happening.